src/librustc/hir/lowering.rs

```diff
@@ -1566,16 +1558,25 @@ pub fn lower_expr(lctx: &LoweringContext, e: &Expr) -> P<hir::Expr> {
                 // to:
                 //
                 // {
-                //     match <expr> {
+                //     match { Carrier::translate( { <expr> } ) } {
```

Comment:
For the record the idea that any error `Carrier`/`ResultCarrier` type should be *isomorphic* to `Result` was very intentional. (Not suggesting that this is therefore beyond debate. Just that it wasn't an oversight.)